### ISSUE TITLE
Add Type Hints/Annotations (PEP-484)

### ISSUE THREAD

**Issue by ciarancourtney** [buildbot/buildbot#4780]
As Python 2 support has been dropped in buildbot-master v2.0, it is now a good opportunity to add [Type Hints](https://docs.python.org/3/library/typing.html). My ideal workflow in an IDE like PyCharm is to instantiate classes directly, read docstrings for arg options (missing in many cases) and have type checked, intellisense-ed&trade; code completion. I think this would be a massively useful addition.

[MonkeyType](https://github.com/Instagram/MonkeyType) could be used to generate initial types from integration tests perhaps.

Keywords: static type typing checking type-checked pep484

**Comment by p12tic** [comment]
Thanks for raising this idea. I agree that it would be an improvement on the codebase to use type hints. Unfortunately, the API is not yet stable, so we can't use it. The following is an excerpt from the docs:

> The typing module has been included in the standard library on a provisional basis. New features might be added and API may change even between minor releases if deemed necessary by the core developers.

Thus I'm closing the issue for now. Once the API stabilizes and there are known best practices we can reopen this.

**Comment by terencehonles** [comment]
@p12tic typing hasn't been provisional since 3.7 and while I understand 3.6 is still supported typing isn't going to change drastically. Should we / can we reopen this issue?

**Comment by tardyp** [comment]
in my experience with vscode, typed based intellisense has been working quite well and has been useful.

I would be in favor of adding simple type hints when there is no polymorphism.

When there is, i.e. a function is accepting several types for one argument, then I would refrain from using it.

That would make union typing necessary and I find that those adds a lot of weight in the code, for quite few benefit.

I have been studying the source code of graphql-core lately, and I find it quite difficult to read, because of typing:

https://github.com/graphql-python/graphql-core/blob/main/src/graphql/execution/execute.py

65 lines of type imports, nearly all the function there require multiline signature. 
I would like to avoid that.

Another example of type definition, that imho does not add much to have (because the type system is not quite ready yet):
https://github.com/graphql-python/graphql-core/blob/main/src/graphql/type/definition.py#L583

**Comment by terencehonles** [comment]
I agree, the more dynamic the more cumbersome and harder it is to type things in a brief manner, but sometimes the extra line weight does make it easier to understand what the method in question expects. I would suggest avoid creating aliases to Any or untyped generics i.e. https://github.com/graphql-python/graphql-core/blob/03c2b06e4636ed8a89c7a4c7e56d244fd9d00bde/src/graphql/type/definition.py#L586 since mypy (I don't know about other checkers) will not preserve the original type so the only thing it adds is a "unique" name for it in one's codebase. It does allow not having to touch all the places it was used *if* and *only if* the existing code conforms to the type constraints one adds in the future, but my opinion is you can just increase the type safety then instead of "tricking one's self" into thinking the code is safer than you had thought.
